charms/knative-operator/src/charm.py

```diff
@@ -0,0 +1,174 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+import traceback
+
+from ops.main import main
+from ops.pebble import Layer
+from ops.charm import CharmBase
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus, StatusBase
+from jinja2 import Environment, FileSystemLoader
+from lightkube import ApiError, Client, codecs
+from serialized_data_interface import NoCompatibleVersions, NoVersionsListed, get_interfaces
+
+logger = logging.getLogger(__name__)
+
+
+class KnativeOperatorCharm(CharmBase):
+    """A Juju Charm for Training Operator"""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self.logger = logging.getLogger(__name__)
+        self.interfaces = ""
+
+        self._name = self.model.app.name
+        self._namespace = self.model.name
+        self._operator_service = "/ko-app/operator"
+        self.env = Environment(loader=FileSystemLoader('src'))
+        self._container = self.unit.get_container(self._name)
+        self._resources_files = {
+             "auth_manifests.yaml",
+             "service_account.yaml",
+             "crds_manifests.yaml",
+             # Skipping config manifests for now
+             # "config_manifests.yaml",
+        }
+        self._context = {"namespace": self._namespace, "name": self._name}
+
+        self.lightkube_client = Client(namespace=self._namespace, field_manager="lightkube")
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.config_changed, self._on_install)
+        self.framework.observe(
+            self.on.knative_operator_pebble_ready,
+            self._on_knative_operator_pebble_ready,
+        )
+        self.framework.observe(self.on["serving"].relation_changed, self._serving_info)
+        self.framework.observe(self.on["eventing"].relation_changed, self._eventing_info)
+
```

Comment:
Not so much a comment about the code but what's not here - we should have something handling the `update-status` event that can check whether we're still alive.  But not as necessary for a first release